File: scraper/functions/collect/dealls.ts
import { DEALLS_API, DEALLS_HEADERS } from '../../core/config';

interface DeallsDoc {
  id: string;
  slug: string;
  role: string;
  employmentTypes: string[];
  workplaceType: string;
  publishedAt: string;
  salaryRange: { start: number; end: number } | null;
  salaryType: string;
  country: { name: string; id: number };
  city: { name: string; id: number } | null;
  company: {
    name: string;
    logoUrl: string | null;
    sector: string;
    verified: boolean;
  };
  skills: { name: string; id: string }[];
  urgentlyNeeded: boolean;
  url?: string;
}

interface DeallsResponse {
  data: {
    docs: DeallsDoc[];
    totalDocs: number;
    totalPages: number;
    page: number;
  };
}

async function fetchPage(page: number, limit: number = 18): Promise<DeallsResponse> {
  const params = new URLSearchParams({
    page: String(page),
    sortParam: 'mostRelevant',
    sortBy: 'asc',
    boostTheBoostedJob: 'true',
    published: 'true',
    limit: String(limit),
    status: 'active',
  });

  const res = await fetch(`${DEALLS_API}?${params}`, {
    method: 'GET',
    headers: DEALLS_HEADERS,
  });

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Dealls API error: ${res.status} ${res.statusText} — ${body.slice(0, 500)}`);
  }

  return res.json();
}

export async function collectDeallsJobs(maxPages: number = 5): Promise<DeallsDoc[]> {
  const all: DeallsDoc[] = [];
  const first = await fetchPage(1, 18);
  all.push(...first.data.docs);

  const totalPages = Math.min(maxPages, first.data.totalPages);
  for (let p = 2; p <= totalPages; p++) {
    const page = await fetchPage(p, 18);
    all.push(...page.data.docs);
  }

  return all;
}

File: scraper/dealls.py
"""Collect job listings from the Dealls (sejutacita) REST API."""
import json
import urllib.parse
import urllib.request

from config import DEALLS_API, DEALLS_HEADERS, REQUEST_TIMEOUT


def _fetch_page(page, limit=18):
    params = urllib.parse.urlencode({
        "page": page,
        "sortParam": "mostRelevant",
        "sortBy": "asc",
        "boostTheBoostedJob": "true",
        "published": "true",
        "limit": limit,
        "status": "active",
    })

    req = urllib.request.Request(f"{DEALLS_API}?{params}", headers=DEALLS_HEADERS, method="GET")
    with urllib.request.urlopen(req, timeout=REQUEST_TIMEOUT) as res:
        return json.loads(res.read().decode("utf-8"))


def collect_jobs(max_pages=5, limit=18):
    first = _fetch_page(1, limit)
    docs = list(first["data"]["docs"])

    total_pages = min(max_pages, first["data"]["totalPages"])
    for page in range(2, total_pages + 1):
        result = _fetch_page(page, limit)
        docs.extend(result["data"]["docs"])

    return docs
